### PR TITLE
Templates

### PR DESCRIPTION
**Pull request by nics**
hi @netsensei this is an improved renderer, can you take a look before I merge?

- partials start with an underscore and are all loaded in a shared template like before
- layouts have a '.layout.gohtml' extension and are all loaded in a shared template
- views each have their own template that is cloned from the layouts template

this is a compromise with the version I showed you where views had access to all partials; that worked fine, but memory exploded from 20mb to 80mb

this version gives the same functionality but memory needed remains about the same
only small downside is that "template" directives are replaced by "partial" because views have no direct access to the partial definitions, so a missing partial will only be detected at runtime

most of the relevant code is in [internal/render/renderer.go](https://github.com/ugent-library/biblio-backend/pull/443/files#diff-9b6a687f37d66fc95d8203999060d17efcfc894d09d0ff40015a6e9879ca406d)

### CODE REVIEW

**Comment by netsensei** [comment]
I think we need to review the structuring of the `views` directory. It's a bit confusing.

* It's called `views` but at the root level, it does contain partials e.g. `_close_modal.gohtml, `_flash.gohtml`.
* Layouts are in their own `layout` folder at the root level. But partials and views share the same folder, only differentiating them with a underscore.
* There are `pages` directories e.g. `dataset/pages` and `pages` and `publication/pages` => it's unclear what the semantics of those are.
* I can spot at least 3 types of views: modal dialogs (e.g. add_publication.gohtml), tabs (e.g. show_publications.gohtml) and refreshes (e.g. refresh_abstract.go) in dataset and publication directories. But they are all lugged into the `dataset` or `publication` directory; whereas on the root level, there's `form`, `fields` and `displays`.  The organization of the templates is subdivided into sets which are named in a fashion which is too generic, or too specific.

So, the goal would be to establish a handful clear conventions that help organize template code.

In that regard, the renderer doesn't care about the organization / naming of templates, so there's lots of flexibility here. Neither should it enforce particular conventions in a hard way as there will be unavoidable exceptions.

File: internal/render/renderer.go [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+func (r *Renderer) Parse() (*Renderer, error) {
+	var (
+		layoutsTemplate  = template.New("")
+		partialsTemplate = template.New("")
+		viewTemplates    = map[string]*template.Template{}
+	)
+
+	for _, funcs := range r.funcMaps {
+		layoutsTemplate.Funcs(funcs)
+		partialsTemplate.Funcs(funcs)
```
[the remark this answers is not in the document]
Comment (nics) [reply internal/render/renderer.go:129]: it's not nonsensical, layout are just html wrappers, the're not only for pages
e.g. calling an alert or a modal layout from a partial makes senses

File: internal/render/renderer.go [hunk 2]
```diff
@@ -0,0 +1,358 @@
...
+	"os"
+	"path"
+	"path/filepath"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+)
+
+var (
```
[the remark this answers is not in the document]
Comment (nics) [reply internal/render/renderer.go:20]: they're global defaults that get used if you don't provide a value to the instance
if you don't call r.Dir('mytemplates/') for example, then the default dir is used

File: internal/render/renderer.go [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+		viewTemplates    = map[string]*template.Template{}
+	)
+
+	for _, funcs := range r.funcMaps {
+		layoutsTemplate.Funcs(funcs)
+		partialsTemplate.Funcs(funcs)
+	}
+
+	// parse layouts
+	err := filepath.Walk(r.dir, func(f string, info os.FileInfo, err error) error {
```
[the remark this answers is not in the document]
Comment (nics) [reply internal/render/renderer.go]: yes, changed it

**Pull request merged**